### PR TITLE
V2 Foundation: Groups, Middlewares, Simplified Interfaces, ...

### PR DESCRIPTION
**Pull request by shomali11**
Removed dead code
Removed plenty of customization code
Simplified Interfaces
Added support for Groups, Middleware
Enhanced support for Interaction Callbacks
Enhanced support for Cron Jobs
Added support for more message options
Improved documentation

### CODE REVIEW

**Comment by arusso** [comment]
In light of #135 the dead code was in fact working at one point and it was bypassed in a regression introduced in d3d36ce8fb8cf7a9110034b02acf671ad6a37146. I'll fix that regression in a separate PR, and that code removal should come out of this PR.

**Comment by arusso** [comment]
I've started a review, but it'll take me time to do a deeper dive into things. There's a lot of changes in a single commit that makes this hard to follow.

I know you've mentioned how you wanted to rid yourself of the customization code, but It's not clear to me _why_ we're removing so much of the customization commands when they were features users requested and likely being used/depended on.

I'll comment more in inline in my review, but a quick example is #97 -- if we're removing the custom command constructor, do we have another way to support this now? Based on #69 it seems like we can do some of this (like specify a custom prefix) in the Command definition, but I'd hesitate to pull this out without fully understanding how this is being used and potentially breaking and having to reimplement this.

**Comment by shomali11** [comment]
I brought back the BotMode code.

Once you review and approve, I will merge squash with the correct commit message requirements.

File: slacker.go [hunk 1]
```diff
@@ -115,11 +102,6 @@ func (s *Slacker) Init(initHandler func()) {
 	s.initHandler = initHandler
 }
 
-// Err handle when errors are encountered
-func (s *Slacker) Err(errorHandler func(err string)) {
-	s.errorHandler = errorHandler
-}
-
```
Comment (arusso) [inline slacker.go:122]: Looks unused, ok to go.

File: slacker.go [hunk 2]
```diff
@@ -130,36 +112,6 @@ func (s *Slacker) Interactive(interactiveEventHandler func(InteractiveBotContext
 	s.interactiveEventHandler = interactiveEventHandler
 }
 
-// CustomBotContext creates a new bot context
-func (s *Slacker) CustomBotContext(botContextConstructor func(context.Context, *slack.Client, *socketmode.Client, *MessageEvent) BotContext) {
-	s.botContextConstructor = botContextConstructor
-}
```
Comment (arusso) [inline slacker.go:136]: This breaks #91 / #103. I don't think we have enough data as to how this is used to say if we provide another means to do what they want without it.

File: slacker.go [hunk 3]
```diff
@@ -130,36 +112,6 @@ func (s *Slacker) Interactive(interactiveEventHandler func(InteractiveBotContext
...
-
-// CustomInteractiveBotContext creates a new interactive bot context
-func (s *Slacker) CustomInteractiveBotContext(interactiveBotContextConstructor func(context.Context, *slack.Client, *socketmode.Client, *socketmode.Event) InteractiveBotContext) {
-	s.interactiveBotContextConstructor = interactiveBotContextConstructor
-}
```
Comment (arusso) [inline slacker.go:141]: Not clear to me why this was needed to begin with and recently introduced so im ok removing.

File: slacker.go [hunk 4]
```diff
@@ -130,36 +112,6 @@ func (s *Slacker) Interactive(interactiveEventHandler func(InteractiveBotContext
...
-
-// CustomJobContext creates a new job context
-func (s *Slacker) CustomJobContext(jobContextConstructor func(context.Context, *slack.Client, *socketmode.Client) JobContext) {
-	s.jobContextConstructor = jobContextConstructor
-}
-
-// CustomCommand creates a new BotCommand
-func (s *Slacker) CustomCommand(commandConstructor func(usage string, definition *CommandDefinition) Command) {
-	s.commandConstructor = commandConstructor
-}
```
Comment (arusso) [inline slacker.go:151]: This breaks https://github.com/shomali11/slacker/pull/91 / https://github.com/shomali11/slacker/pull/103. I don't think we have enough data as to how this is used to say if we provide another means to do what they want without it.

File: slacker.go [hunk 5]
```diff
@@ -130,36 +112,6 @@ func (s *Slacker) Interactive(interactiveEventHandler func(InteractiveBotContext
...
-// CustomRequest creates a new request
-func (s *Slacker) CustomRequest(requestConstructor func(botCtx BotContext, properties *proper.Properties) Request) {
-	s.requestConstructor = requestConstructor
-}
-
-// CustomResponse creates a new response writer
-func (s *Slacker) CustomResponse(responseConstructor func(botCtx BotContext) ResponseWriter) {
-	s.responseConstructor = responseConstructor
-}
-
```
Comment (arusso) [inline slacker.go:162]: What's replacing it? Do we have an example on how to do this without these?

File: context.go [hunk 6]
```diff
@@ -3,99 +3,155 @@ package slacker
...
 	"github.com/slack-go/slack"
 	"github.com/slack-go/slack/socketmode"
 )
 
-// BotContext interface is for bot command contexts
-type BotContext interface {
+// CommandContext interface is for bot command contexts
+type CommandContext interface {
 	Context() context.Context
+	Usage() string
```
Comment (arusso) [inline context.go:14]: Why is this on the Context and not the Definition?

**Review by arusso — COMMENTED** [review]
I really where this is going -- UX definitely seems to be improved for sure and the Gin-style groups and middleware should will resolve some of my biggest annoyances.

For logging I think it would be better to have a small `Logger` interface that defines `Debugf()` and `Infof()` which users can provide their own implementation / format for. This can probably be its own PR since its well bounded. I wouldn't mind taking a stab at this once this PR merges in if you don't want to do it.

Some fairly minor things:
- I don't see a good reason to keep calling it `botCtx`, `ctx` is both shorter and idiomatic
- I believe since Go 1.18 using `any` in favor of the empty interface `interface{}` is preferred

I'll find some time to exercise these changes and may have additional feedback once I do.

File: examples/command-defaults/main.go
Comment (arusso) [inline examples/command-defaults/main.go:1]: We should combine this into the command-parameters example so there is one spot to look for all things command parameters.

File: examples/slash-command/main.go
Comment (arusso) [inline examples/slash-command/main.go:1]: Seems like this could be rolled into the basic command example with a well placed comment.

File: go.sum
Comment (arusso) [inline go.sum:1]: Can this be pruned further with `go mod tidy`?

File: message_event.go [hunk 7]
```diff
@@ -60,20 +60,20 @@ func (e *MessageEvent) IsThread() bool {
 
 // IsBot indicates if the message was sent by a bot
 func (e *MessageEvent) IsBot() bool {
-	return e.BotID != ""
```
Comment (arusso) [inline message_event.go:63]: This seems like an unnecessary stylistic change and less clear imo.

`MessageEvent.BotID` is a string and it seems clearer to treat it as such, since we don't normally interact with them by index like we might a slice where checking length would be more appropriate.

At the end of the day they should both compile to the [same thing](https://godbolt.org/z/631ncd9ns).

File: response.go [hunk 8]
```diff
@@ -1,85 +1,98 @@
...
-	"log"
-
 	"github.com/slack-go/slack"
 )
 
-const (
-	errorFormat = "*Error:* _%s_"
-)
+// WriterReplierResponse interface is used to respond to an event
+type WriterReplierResponse interface {
```
Comment (arusso) [inline response.go:8]: Needs docs, and does the interface give us any benefit here? If it's needed, why not just embed the interfaces into a super interface like:

```
type WriterReplierResponse interface {
  Replier
  WriterResponse
}
```

File: response.go [hunk 9]
```diff
@@ -1,85 +1,98 @@
...
-	if ev == nil {
-		return fmt.Errorf("unable to get message event details")
-	}
-	return r.Post(ev.ChannelID, message, options...)
+// Delete deletes a message in a channel
+func (r *writerReplierResponse) Delete(channel string, messageTimestamp string) (string, error) {
+	return r.poster.Delete(channel, messageTimestamp)
+}
+
+// WriterResponse interface is used to respond to an event
```
Comment (arusso) [inline response.go:63]: Not sure I like the name of this. Maybe `PostDeleter` (like `io.ReadWriter`) or a less confusing `PosterDeleter`?

File: response_writer.go [hunk 10]
```diff
@@ -0,0 +1,97 @@
...
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/slack-go/slack"
+)
+
+// Writer interface is used to respond to an event
+type Writer interface {
```
Comment (arusso) [inline response_writer.go:11]: Why duplicate `WriterResponse` interface? As an interface name, doesn't really line up with what it does. See my comments on `WriterResponse` interface about naming.

**Review by arusso — COMMENTED** [review]
I like where this is going, though I feel like we might be using interfaces a bit more than needed. Where they are needed they could be beneficial we're duplicating a lot where we could be composing multiple smaller interfaces into larger ones and benefitting from consistent signatures in all definitions.

I was hoping logging would provide users the ability to pass in a structured logger in place of a normal logger, but that doesn't seem possible here.

While not a problem we need to fix in the initial release, Jobs makes the assumption that we're running a single instance of a bot -- which since SocketMode introduce built-in HA, is likely not to be ubiquitous. We should make a not and work on a locking interface to let people pick redis/consul/locking-db-of-choice to lock jobs to a single instance.

I'd like to implement a bot using what we've got, and I'll likely have more input after I do that.

**Comment by arusso** [comment]
The CI errors can be fixed by squashing everything down into a single commit and properly formatting that commit subject.

File: command_group.go [hunk 11]
```diff
@@ -0,0 +1,59 @@
+package slacker
+
+import (
+	"fmt"
+	"strings"
+)
+
+// CommandGroup a group of commands
```
[the remark this answers is not in the document]
Comment (shomali11) [reply command_group.go:8]: Good point. I will get rid of the interface.

File: context.go [hunk 12]
```diff
@@ -3,119 +3,188 @@ package slacker
...
 
+	"github.com/shomali11/proper"
 	"github.com/slack-go/slack"
-	"github.com/slack-go/slack/socketmode"
 )
 
-// BotContext interface is for bot command contexts
-type BotContext interface {
+// CommandContext interface is for bot command contexts
+type CommandContext interface {
```
[the remark this answers is not in the document]
Comment (shomali11) [reply context.go:11]: Will remove it

File: examples/command-defaults/main.go
Comment (shomali11) [reply examples/command-defaults/main.go:1]: Sounds good

File: examples/slash-command/main.go
Comment (shomali11) [reply examples/slash-command/main.go:1]: Done

File: go.sum
Comment (shomali11) [reply go.sum:1]: Done

File: message_event.go (hunk 7 above)
Comment (shomali11) [reply message_event.go:63]: Interesting. I remember `len(stringValue) > 0` performing much much better than `stringValue != ""`

No worries, I can revert it

File: response.go (hunk 8 above)
Comment (shomali11) [reply response.go]: Cleaned it up

File: response.go (hunk 9 above)
Comment (shomali11) [reply response.go]: Renamed it

File: response_writer.go (hunk 10 above)
Comment (shomali11) [reply response_writer.go]: Got rid of the interfaces

File: app_manifest/manifest.yml
[the remark this answers is not in the document]
Comment (shomali11) [reply app_manifest/manifest.yml:1]: I felt that moving it to a different directory made sense in case we wanted to create several manifests for various different cases.

Plus, I did not want it to get lost among the code examples
